File: packages/ld-adapter/src/__tests__/semantic-patch-builder.test.ts
import { describe, expect, it } from 'vitest';
import { buildRolloutWeights, buildTargetingPatchBody } from '../write/semantic-patch.js';

describe('semantic patch builders', () => {
	it('buildRolloutWeights returns weights summing to 100000', () => {
		const weights = buildRolloutWeights(10_000, 'var-on', 'var-off');
		expect(weights).toEqual({ 'var-on': 10_000, 'var-off': 90_000 });
		expect(Object.values(weights).reduce((a, b) => a + b, 0)).toBe(100_000);
	});

	it('buildTargetingPatchBody for fallthrough emits turnFlagOn + updateFallthroughVariationOrRollout', () => {
		const body = buildTargetingPatchBody(
			{
				environmentKey: 'production',
				rollout: {
					mode: 'fallthrough',
					treatmentVariationRef: { by: 'id', id: 'var-on' },
					controlVariationRef: { by: 'id', id: 'var-off' },
					treatmentPercentThousandths: 10_000,
					rolloutContextKind: 'user',
					rolloutBucketBy: 'user',
				},
			},
			{
				treatmentVariationId: 'var-on',
				controlVariationId: 'var-off',
			},
		);

		expect(body.instructions.map((i) => i.kind)).toEqual([
			'turnFlagOn',
			'updateFallthroughVariationOrRollout',
		]);
		expect(body.instructions[1]).toMatchObject({
			rolloutBucketBy: 'user',
			rolloutContextKind: 'user',
		});
	});

	it('never includes updatePercentageRollout instruction kind', () => {
		const body = buildTargetingPatchBody(
			{
				environmentKey: 'production',
				rollout: {
					mode: 'fallthrough',
					treatmentVariationRef: { by: 'id', id: 'a' },
					controlVariationRef: { by: 'id', id: 'b' },
					treatmentPercentThousandths: 50_000,
					rolloutContextKind: 'user',
					rolloutBucketBy: 'user',
				},
			},
			{ treatmentVariationId: 'a', controlVariationId: 'b' },
		);
		for (const instruction of body.instructions) {
			expect(instruction.kind).not.toBe('updatePercentageRollout');
		}
	});
});

File: kotlin/modules/ld-adapter/src/test/kotlin/com/ffpromo/ldadapter/write/SemanticPatchBuilderTest.kt
package com.ffpromo.ldadapter.write

import com.ffpromo.contracts.RolloutIntent
import com.ffpromo.contracts.TargetingIntent
import com.ffpromo.contracts.VariationRef
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class SemanticPatchBuilderTest {
    @Test
    fun `buildRolloutWeights returns weights summing to 100000`() {
        val weights = buildRolloutWeights(10_000, "var-on", "var-off")
        assertEquals(mapOf("var-on" to 10_000, "var-off" to 90_000), weights)
        assertEquals(100_000, weights.values.sum())
    }

    @Test
    fun `buildTargetingPatchBody for fallthrough emits turnFlagOn and updateFallthroughVariationOrRollout`() {
        val body = buildTargetingPatchBody(
            TargetingIntent(
                environmentKey = "production",
                rollout = RolloutIntent(
                    mode = "fallthrough",
                    treatmentVariationRef = VariationRef.ById("var-on"),
                    controlVariationRef = VariationRef.ById("var-off"),
                    treatmentPercentThousandths = 10_000,
                    rolloutContextKind = "user",
                    rolloutBucketBy = "user",
                ),
            ),
            ResolvedRolloutIds(
                treatmentVariationId = "var-on",
                controlVariationId = "var-off",
            ),
        )

        val kinds = body["instructions"]!!.toString()
        assert(kinds.contains("turnFlagOn"))
        assert(kinds.contains("updateFallthroughVariationOrRollout"))
        assert(kinds.contains("rolloutBucketBy"))
        assert(kinds.contains("user"))
    }

    @Test
    fun `never includes updatePercentageRollout instruction kind`() {
        val body = buildTargetingPatchBody(
            TargetingIntent(
                environmentKey = "production",
                rollout = RolloutIntent(
                    mode = "fallthrough",
                    treatmentVariationRef = VariationRef.ById("a"),
                    controlVariationRef = VariationRef.ById("b"),
                    treatmentPercentThousandths = 50_000,
                    rolloutContextKind = "user",
                    rolloutBucketBy = "user",
                ),
            ),
            ResolvedRolloutIds(treatmentVariationId = "a", controlVariationId = "b"),
        )

        assertFalse(body.toString().contains("updatePercentageRollout"))
    }
}
